File: src/parser.ts
/**
 * @file parser.ts
 * @description HTML 解析器 - 从上海图书馆网页中提取结构化数据
 * 
 * 本文件负责将上海图书馆 VuFind 系统返回的 HTML 页面解析为结构化的 JSON 数据。
 * 
 * 【项目地位】
 * - 作为数据处理层的核心组件，是数据获取与业务逻辑之间的桥梁
 * - 被 library-client.ts 和 cli.js 共同使用
 * 
 * 【主要功能】
 * 1. 搜索结果解析：从搜索结果页面提取图书列表和分页信息
 * 2. 图书详情解析：从图书详情页面提取完整元数据
 * 3. 馆藏信息解析：从馆藏页面提取各分馆的借阅状态
 * 
 * 【解析策略】
 * - 使用正则表达式匹配 HTML 结构
 * - 支持中英文两种页面格式
 * - 自动解码 HTML 实体
 * 
 * @author ZedeX
 * @version 1.0.0
 * @date 2026-04-18
 * @license Apache-2.0
 */

/**
 * 解析后的图书信息接口（搜索结果）
 * @interface ParsedBook
 */
export interface ParsedBook {
  record_id: string;
  title: string;
  author: string;
  publisher: string;
  publish_year: string;
  call_number: string;
  cover_url: string;
  availability_summary: string;
}

/**
 * 解析后的搜索结果接口
 * @interface ParsedSearchResult
 */
export interface ParsedSearchResult {
  books: ParsedBook[];
  total_results: number;
  total_pages: number;
  current_page: number;
}

/**
 * 解析后的图书详情接口
 * @interface ParsedBookDetail
 */
export interface ParsedBookDetail {
  record_id: string;
  title: string;
  author: string;
  publisher: string;
  publish_year: string;
  call_number: string;
  cover_url: string;
  isbn: string;
  summary: string;
}

/**
 * 解析后的馆藏信息接口
 * @interface ParsedHolding
 */
export interface ParsedHolding {
  library: string;
  location: string;
  call_number: string;
  status: string;
  record_id: string;
}

/**
 * 解码 HTML 实体
 * 将 HTML 实体字符转换为对应的 Unicode 字符
 * @function decodeHtmlEntities
 * @param {string} text - 包含 HTML 实体的文本
 * @returns {string} 解码后的文本
 */
function decodeHtmlEntities(text: string): string {
  return text
    .replace(/&amp;/g, '&')
    .replace(/&lt;/g, '<')
    .replace(/&gt;/g, '>')
    .replace(/&quot;/g, '"')
    .replace(/&#39;/g, "'")
    .replace(/&nbsp;/g, ' ')
    .replace(/&#x2F;/g, '/')
    .replace(/&#x3B;/g, ';')
    .replace(/&#x3D;/g, '=')
    .replace(/&#x20;/g, ' ');
}

/**
 * 移除 HTML 标签
 * 将 HTML 字符串转换为纯文本
 * @function stripTags
 * @param {string} html - HTML 字符串
 * @returns {string} 纯文本字符串
 */
function stripTags(html: string): string {
  return html.replace(/<[^>]+>/g, ' ').replace(/\s+/g, ' ').trim();
}

/**
 * 解析搜索结果页面
 * 从上海图书馆搜索结果 HTML 页面中提取图书列表和分页信息
 * @function parseSearchResults
 * @param {string} html - 搜索结果页面的 HTML 内容
 * @returns {ParsedSearchResult} 解析结果对象
 */
export function parseSearchResults(html: string): ParsedSearchResult {
  const books: ParsedBook[] = [];
  let totalResults = 0;
  let totalPages = 1;
  let currentPage = 1;
  let perPage = 20;

  const totalMatch = html.match(/共\s*(\d[\d,]*)\s*条/);
  if (totalMatch) {
    totalResults = parseInt(totalMatch[1].replace(/,/g, ''), 10);
  }

  const showingMatch = html.match(/Showing\s+<strong>(\d+)\s*-\s*(\d+)<\/strong>/i)
    || html.match(/Showing[^<]*<strong>(\d+)\s*-\s*(\d+)<\/strong>/i);
  if (showingMatch) {
    perPage = parseInt(showingMatch[2], 10) - parseInt(showingMatch[1], 10) + 1;
  }

  const cnRangeMatch = html.match(/第\s*<strong>(\d+)\s*-\s*(\d+)\s*条<\/strong>/);
  if (cnRangeMatch) {
    perPage = parseInt(cnRangeMatch[2], 10) - parseInt(cnRangeMatch[1], 10) + 1;
  }

  const pageMatches = [...html.matchAll(/page=(\d+)/g)];
  if (pageMatches.length > 0) {
    const pageNums = pageMatches.map(m => parseInt(m[1], 10)).filter(n => n > 0);
    if (pageNums.length > 0) {
      totalPages = Math.max(...pageNums);
    }
  }

  if (totalResults === 0 && totalPages > 0) {
    totalResults = totalPages * perPage;
  }

  const activePageMatch = html.match(/class="active"[^>]*><span>(\d+)<\/span>/);
  if (activePageMatch) {
    currentPage = parseInt(activePageMatch[1], 10);
  } else {
    const urlPageMatch = html.match(/[?&]page=(\d+)/);
    if (urlPageMatch) {
      currentPage = parseInt(urlPageMatch[1], 10);
    }
  }

  const resultBlocks = html.split(/(?=<div[^>]*id="result\d+")/);
  for (const block of resultBlocks) {
    if (!/id="result\d+"/.test(block)) continue;

    const book: ParsedBook = {
      record_id: '',
      title: '',
      author: '',
      publisher: '',
      publish_year: '',
      call_number: '',
      cover_url: '',
      availability_summary: '',
    };

    const recordIdMatch = block.match(/\/Record\/([^"?/]+)/);
    if (recordIdMatch) {
      book.record_id = recordIdMatch[1];
    }

    const titleMatch = block.match(/<a[^>]*class="[^"]*title[^"]*"[^>]*>([\s\S]*?)<\/a>/);
    if (titleMatch) {
      book.title = decodeHtmlEntities(stripTags(titleMatch[1])).trim();
    }

    const authorMatches = [...block.matchAll(/<a[^>]*href="[^"]*type=Author[^"]*"[^>]*>([\s\S]*?)<\/a>/g)];
    if (authorMatches.length > 0) {
      book.author = authorMatches.map(m => decodeHtmlEntities(stripTags(m[1])).trim()).join('; ');
    }

    const coverMatch = block.match(/<img[^>]*src="([^"]*Cover[^"]*)"/);
    if (coverMatch) {
      book.cover_url = decodeHtmlEntities(coverMatch[1]);
    }

    const bodyText = stripTags(block);

    const publisherMatch = bodyText.match(/Published:\s*(.+?)(?:\s+Publication Dates:|\s+Call Number:|\s+查询馆藏|$)/);
    if (publisherMatch) {
      book.publisher = publisherMatch[1].trim();
    } else {
      const pubMatch2 = bodyText.match(/出版社[：:]\s*(.+?)(?:\s+出版时间[：:]|\s+索书号[：:]|\s+查询馆藏|$)/);
      if (pubMatch2) {
        book.publisher = pubMatch2[1].trim();
      }
    }

    const yearMatch = bodyText.match(/Publication Dates:\s*(\d{4})/);
    if (yearMatch) {
      book.publish_year = yearMatch[1];
    } else {
      const yearMatch2 = bodyText.match(/出版时间[：:]\s*(\d{4})/);
      if (yearMatch2) {
        book.publish_year = yearMatch2[1];
      }
    }

    const callNumberMatch = bodyText.match(/Call Number:\s*(\S+)/);
    if (callNumberMatch) {
      book.call_number = callNumberMatch[1];
    } else {
      const cnMatch2 = bodyText.match(/索书号[：:]\s*(\S+)/);
      if (cnMatch2) {
        book.call_number = cnMatch2[1];
      }
    }

    if (book.record_id) {
      books.push(book);
    }
  }

  return { books, total_results: totalResults, total_pages: totalPages, current_page: currentPage };
}

/**
 * 解析图书详情页面
 * 从上海图书馆图书详情 HTML 页面中提取图书详细信息
 * @function parseBookDetail
 * @param {string} html - 图书详情页面的 HTML 内容
 * @param {string} recordId - 图书记录 ID
 * @returns {ParsedBookDetail} 图书详情对象
 */
export function parseBookDetail(html: string, recordId: string): ParsedBookDetail {
  const detail: ParsedBookDetail = {
    record_id: recordId,
    title: '',
    author: '',
    publisher: '',
    publish_year: '',
    call_number: '',
    cover_url: '',
    isbn: '',
    summary: '',
  };

  const titleMatch = html.match(/<h3[^>]*property="name"[^>]*>([\s\S]*?)<\/h3>/);
  if (titleMatch) {
    detail.title = decodeHtmlEntities(stripTags(titleMatch[1]));
  }

  const authorMatches = [...html.matchAll(/<span[^>]*class="[^"]*author-data[^"]*"[^>]*>[\s\S]*?<a[^>]*>([\s\S]*?)<\/a>/g)];
  if (authorMatches.length > 0) {
    detail.author = authorMatches.map(m => decodeHtmlEntities(stripTags(m[1])).trim()).join('; ');
  }

  const publisherMatch = html.match(/<span[^>]*property="publisher"[^>]*>([\s\S]*?)<\/span>/);
  if (publisherMatch) {
    detail.publisher = decodeHtmlEntities(stripTags(publisherMatch[1]));
  }

  const coverMatch = html.match(/<img[^>]*src="([^"]*Cover[^"]*)"/);
  if (coverMatch) {
    detail.cover_url = decodeHtmlEntities(coverMatch[1]);
  }

  const summaryMatch = html.match(/<div[^>]*class="[^"]*summary[^"]*"[^>]*>([\s\S]*?)<\/div>/);
  if (summaryMatch) {
    detail.summary = decodeHtmlEntities(stripTags(summaryMatch[1]));
  }

  const detailTableMatch = html.match(/<table[^>]*id="table-detail"[^>]*>([\s\S]*?)<\/table>/);
  if (detailTableMatch) {
    const tableHtml = detailTableMatch[1];
    const tableRows = tableHtml.matchAll(/<tr[^>]*>([\s\S]*?)<\/tr>/g);
    for (const row of tableRows) {
      const rowHtml = row[1];
      const thMatch = rowHtml.match(/<th[^>]*>([\s\S]*?)<\/th>/);
      const tdMatch = rowHtml.match(/<td[^>]*>([\s\S]*?)<\/td>/);
      if (thMatch && tdMatch) {
        const th = stripTags(thMatch[1]).trim();
        const td = stripTags(tdMatch[1]).trim();

        if (/isbn/i.test(th)) {
          const isbnMatch = td.match(/[\d\-Xx]+/);
          if (isbnMatch) detail.isbn = isbnMatch[0];
        } else if (/出版/.test(th) && /日期|时间|年/.test(th)) {
          const yearMatch = td.match(/(\d{4})/);
          if (yearMatch) detail.publish_year = yearMatch[1];
        } else if (/索书号/.test(th)) {
          detail.call_number = td;
        }
      }
    }
  }

  if (!detail.cover_url) {
    detail.cover_url = `/Cover/Show?instanceId=${recordId}`;
  }

  return detail;
}

/**
 * 解析馆藏信息页面
 * 从上海图书馆馆藏 HTML 页面中提取各分馆的馆藏状态
 * @function parseHoldings
 * @param {string} html - 馆藏页面的 HTML 内容
 * @param {string} recordId - 图书记录 ID
 * @returns {ParsedHolding[]} 馆藏信息列表
 */
export function parseHoldings(html: string, recordId: string): ParsedHolding[] {
  const holdings: ParsedHolding[] = [];
  let currentLibrary = '';

  const libH3Matches = html.matchAll(/<h3[^>]*>([\s\S]*?)<\/h3>/g);
  const libPositions: { pos: number; library: string }[] = [];
  for (const m of libH3Matches) {
    const text = stripTags(m[1]).trim();
    const libMatch = text.match(/所属馆[：:]\s*(.+)/);
    if (libMatch) {
      libPositions.push({ pos: m.index!, library: libMatch[1].trim() });
    } else if (text) {
      libPositions.push({ pos: m.index!, library: text });
    }
  }

  const trMatches = html.matchAll(/<tr[^>]*>([\s\S]*?)<\/tr>/g);
  for (const tr of trMatches) {
    const trHtml = tr[1];
    const trPos = tr.index!;

    for (const lp of libPositions) {
      if (lp.pos < trPos) {
        currentLibrary = lp.library;
      }
    }

    const callNumberMatch = trHtml.match(/<span[^>]*class="[^"]*callnumber[^"]*"[^>]*>([\s\S]*?)<\/span>/);
    const barcodeMatch = trHtml.match(/<span[^>]*class="[^"]*barcode[^"]*"[^>]*>([\s\S]*?)<\/span>/);
    const statusMatch = trHtml.match(/<span[^>]*class="[^"]*availability[^"]*"[^>]*>([\s\S]*?)<\/span>/);

    const callNumber = callNumberMatch ? stripTags(callNumberMatch[1]).trim() : '';
    const barcode = barcodeMatch ? stripTags(barcodeMatch[1]).trim() : '';
    let status = statusMatch ? stripTags(statusMatch[1]).trim() : '';

    if (!status) {
      const statusText = stripTags(trHtml);
      if (/已归还|可借|在馆/.test(statusText)) {
        const sMatch = statusText.match(/(已归还[^\s]*|可借[^\s]*|在馆[^\s]*)/);
        if (sMatch) status = sMatch[1];
      } else if (/借出|外借/.test(statusText)) {
        const sMatch = statusText.match(/(已借出[^\s]*|外借[^\s]*)/);
        if (sMatch) status = sMatch[1];
      } else if (/丢失/.test(statusText)) {
        status = '馆藏丢失';
      }
    }

    if (status.includes('Available')) {
      status = status.replace('Available', '已归还');
    }
    if (status.includes('Loaned out')) {
      status = status.replace('Loaned out', '已借出');
    }
    if (status.includes('Lost')) {
      status = status.replace('Lost', '馆藏丢失');
    }

    if (callNumber || status) {
      holdings.push({
        library: currentLibrary,
        location: barcode,
        call_number: callNumber,
        status: status,
        record_id: recordId,
      });
    }
  }

  return holdings;
}

File: desktop/parser.go
// parser.go - HTML 解析器
// 从上海图书馆 VuFind 系统返回的 HTML 页面中提取结构化数据
// 移植自 TypeScript 版本的 parser.ts
package main

import (
	"fmt"
	"regexp"
	"strconv"
	"strings"
)

// Book 搜索结果中的图书信息
type Book struct {
	RecordID            string `json:"record_id"`
	Title               string `json:"title"`
	Author              string `json:"author"`
	Publisher           string `json:"publisher"`
	PublishYear         string `json:"publish_year"`
	CallNumber          string `json:"call_number"`
	CoverURL            string `json:"cover_url"`
	AvailabilitySummary string `json:"availability_summary"`
}

// SearchResult 搜索结果
type SearchResult struct {
	Books        []Book `json:"books"`
	TotalResults int    `json:"total_results"`
	TotalPages   int    `json:"total_pages"`
	CurrentPage  int    `json:"current_page"`
}

// BookDetail 图书详情
type BookDetail struct {
	RecordID    string `json:"record_id"`
	Title       string `json:"title"`
	Author      string `json:"author"`
	Publisher   string `json:"publisher"`
	PublishYear string `json:"publish_year"`
	CallNumber  string `json:"call_number"`
	CoverURL    string `json:"cover_url"`
	ISBN        string `json:"isbn"`
	Summary     string `json:"summary"`
}

// Holding 馆藏信息
type Holding struct {
	Library   string `json:"library"`
	Location  string `json:"location"`
	CallNumber string `json:"call_number"`
	Status    string `json:"status"`
	RecordID  string `json:"record_id"`
}

// decodeHtmlEntities 解码 HTML 实体
func decodeHtmlEntities(text string) string {
	replacements := []struct {
		entity  string
		decoded string
	}{
		{"&amp;", "&"},
		{"&lt;", "<"},
		{"&gt;", ">"},
		{"&quot;", "\""},
		{"&#39;", "'"},
		{"&nbsp;", " "},
		{"&#x2F;", "/"},
		{"&#x3B;", ";"},
		{"&#x3D;", "="},
		{"&#x20;", " "},
	}
	result := text
	for _, r := range replacements {
		result = strings.ReplaceAll(result, r.entity, r.decoded)
	}
	return result
}

// stripTags 移除 HTML 标签，返回纯文本
func stripTags(html string) string {
	re := regexp.MustCompile(`<[^>]+>`)
	result := re.ReplaceAllString(html, " ")
	// 合并多余空白
	spaceRe := regexp.MustCompile(`\s+`)
	result = spaceRe.ReplaceAllString(result, " ")
	return strings.TrimSpace(result)
}

// ParseSearchResults 解析搜索结果页面
func ParseSearchResults(html string) SearchResult {
	books := []Book{}
	totalResults := 0
	totalPages := 1
	currentPage := 1
	perPage := 20

	// 提取总结果数
	totalRe := regexp.MustCompile(`共\s*(\d[\d,]*)\s*条`)
	if m := totalRe.FindStringSubmatch(html); len(m) > 1 {
		totalResults, _ = strconv.Atoi(strings.ReplaceAll(m[1], ",", ""))
	}

	// 英文格式：Showing <strong>1 - 20</strong>
	showingRe := regexp.MustCompile(`(?i)Showing\s+<strong>(\d+)\s*-\s*(\d+)</strong>`)
	if m := showingRe.FindStringSubmatch(html); len(m) > 2 {
		end, _ := strconv.Atoi(m[2])
		start, _ := strconv.Atoi(m[1])
		perPage = end - start + 1
	}

	// 中文格式：第 <strong>1 - 20</strong> 条
	cnRangeRe := regexp.MustCompile(`第\s*<strong>(\d+)\s*-\s*(\d+)\s*条</strong>`)
	if m := cnRangeRe.FindStringSubmatch(html); len(m) > 2 {
		end, _ := strconv.Atoi(m[2])
		start, _ := strconv.Atoi(m[1])
		perPage = end - start + 1
	}

	// 提取总页数
	pageRe := regexp.MustCompile(`page=(\d+)`)
	pageMatches := pageRe.FindAllStringSubmatch(html, -1)
	if len(pageMatches) > 0 {
		maxPage := 0
		for _, m := range pageMatches {
			if p, err := strconv.Atoi(m[1]); err == nil && p > maxPage {
				maxPage = p
			}
		}
		if maxPage > 0 {
			totalPages = maxPage
		}
	}

	// 如果没有总结果数但有总页数，估算
	if totalResults == 0 && totalPages > 0 {
		totalResults = totalPages * perPage
	}

	// 提取当前页
	activePageRe := regexp.MustCompile(`class="active"[^>]*><span>(\d+)</span>`)
	if m := activePageRe.FindStringSubmatch(html); len(m) > 1 {
		currentPage, _ = strconv.Atoi(m[1])
	} else {
		urlPageRe := regexp.MustCompile(`[?&]page=(\d+)`)
		if m := urlPageRe.FindStringSubmatch(html); len(m) > 1 {
			currentPage, _ = strconv.Atoi(m[1])
		}
	}

	// 按 result div 分割（Go RE2 不支持 (?=) 前瞻，改用 FindAllStringSubmatchIndex 定位）
	resultDivRe := regexp.MustCompile(`<div[^>]*id="result(\d+)"`)
	resultDivMatches := resultDivRe.FindAllStringSubmatchIndex(html, -1)

	for idx, loc := range resultDivMatches {
		// 当前 block 从此匹配开始到下一个匹配开始（或末尾）
		blockStart := loc[0]
		var blockEnd int
		if idx+1 < len(resultDivMatches) {
			blockEnd = resultDivMatches[idx+1][0]
		} else {
			blockEnd = len(html)
		}
		block := html[blockStart:blockEnd]

		book := Book{}

		// 提取记录 ID
		recordIdRe := regexp.MustCompile(`/Record/([^"?/]+)`)
		if m := recordIdRe.FindStringSubmatch(block); len(m) > 1 {
			book.RecordID = m[1]
		}

		// 提取标题
		titleRe := regexp.MustCompile(`<a[^>]*class="[^"]*title[^"]*"[^>]*>([\s\S]*?)</a>`)
		if m := titleRe.FindStringSubmatch(block); len(m) > 1 {
			book.Title = strings.TrimSpace(decodeHtmlEntities(stripTags(m[1])))
		}

		// 提取作者
		authorRe := regexp.MustCompile(`<a[^>]*href="[^"]*type=Author[^"]*"[^>]*>([\s\S]*?)</a>`)
		authorMatches := authorRe.FindAllStringSubmatch(block, -1)
		if len(authorMatches) > 0 {
			authors := []string{}
			for _, m := range authorMatches {
				authors = append(authors, strings.TrimSpace(decodeHtmlEntities(stripTags(m[1]))))
			}
			book.Author = strings.Join(authors, "; ")
		}

		// 提取封面
		coverRe := regexp.MustCompile(`<img[^>]*src="([^"]*Cover[^"]*)"`)
		if m := coverRe.FindStringSubmatch(block); len(m) > 1 {
			book.CoverURL = decodeHtmlEntities(m[1])
		}

		// 提取纯文本用于解析出版信息
		bodyText := stripTags(block)

		// 提取出版社（英文格式）
		pubEnRe := regexp.MustCompile(`Published:\s*(.+?)(?:\s+Publication Dates:|\s+Call Number:|\s+查询馆藏|$)`)
		if m := pubEnRe.FindStringSubmatch(bodyText); len(m) > 1 {
			book.Publisher = strings.TrimSpace(m[1])
		} else {
			// 中文格式
			pubCnRe := regexp.MustCompile(`出版社[：:]\s*(.+?)(?:\s+出版时间[：:]|\s+索书号[：:]|\s+查询馆藏|$)`)
			if m := pubCnRe.FindStringSubmatch(bodyText); len(m) > 1 {
				book.Publisher = strings.TrimSpace(m[1])
			}
		}

		// 提取出版年（英文格式）
		yearEnRe := regexp.MustCompile(`Publication Dates:\s*(\d{4})`)
		if m := yearEnRe.FindStringSubmatch(bodyText); len(m) > 1 {
			book.PublishYear = m[1]
		} else {
			// 中文格式
			yearCnRe := regexp.MustCompile(`出版时间[：:]\s*(\d{4})`)
			if m := yearCnRe.FindStringSubmatch(bodyText); len(m) > 1 {
				book.PublishYear = m[1]
			}
		}

		// 提取索书号（英文格式）
		callEnRe := regexp.MustCompile(`Call Number:\s*(\S+)`)
		if m := callEnRe.FindStringSubmatch(bodyText); len(m) > 1 {
			book.CallNumber = m[1]
		} else {
			// 中文格式
			callCnRe := regexp.MustCompile(`索书号[：:]\s*(\S+)`)
			if m := callCnRe.FindStringSubmatch(bodyText); len(m) > 1 {
				book.CallNumber = m[1]
			}
		}

		if book.RecordID != "" {
			books = append(books, book)
		}
	}

	return SearchResult{
		Books:        books,
		TotalResults: totalResults,
		TotalPages:   totalPages,
		CurrentPage:  currentPage,
	}
}

// ParseBookDetail 解析图书详情页面
func ParseBookDetail(html string, recordId string) BookDetail {
	detail := BookDetail{
		RecordID: recordId,
	}

	// 提取标题
	titleRe := regexp.MustCompile(`<h3[^>]*property="name"[^>]*>([\s\S]*?)</h3>`)
	if m := titleRe.FindStringSubmatch(html); len(m) > 1 {
		detail.Title = decodeHtmlEntities(stripTags(m[1]))
	}

	// 提取作者
	authorRe := regexp.MustCompile(`<span[^>]*class="[^"]*author-data[^"]*"[^>]*>[\s\S]*?<a[^>]*>([\s\S]*?)</a>`)
	authorMatches := authorRe.FindAllStringSubmatch(html, -1)
	if len(authorMatches) > 0 {
		authors := []string{}
		for _, m := range authorMatches {
			authors = append(authors, strings.TrimSpace(decodeHtmlEntities(stripTags(m[1]))))
		}
		detail.Author = strings.Join(authors, "; ")
	}

	// 提取出版社
	pubRe := regexp.MustCompile(`<span[^>]*property="publisher"[^>]*>([\s\S]*?)</span>`)
	if m := pubRe.FindStringSubmatch(html); len(m) > 1 {
		detail.Publisher = decodeHtmlEntities(stripTags(m[1]))
	}

	// 提取封面
	coverRe := regexp.MustCompile(`<img[^>]*src="([^"]*Cover[^"]*)"`)
	if m := coverRe.FindStringSubmatch(html); len(m) > 1 {
		detail.CoverURL = decodeHtmlEntities(m[1])
	}

	// 提取摘要
	summaryRe := regexp.MustCompile(`<div[^>]*class="[^"]*summary[^"]*"[^>]*>([\s\S]*?)</div>`)
	if m := summaryRe.FindStringSubmatch(html); len(m) > 1 {
		detail.Summary = decodeHtmlEntities(stripTags(m[1]))
	}

	// 解析详情表格
	tableRe := regexp.MustCompile(`<table[^>]*id="table-detail"[^>]*>([\s\S]*?)</table>`)
	if m := tableRe.FindStringSubmatch(html); len(m) > 1 {
		tableHtml := m[1]
		rowRe := regexp.MustCompile(`<tr[^>]*>([\s\S]*?)</tr>`)
		rows := rowRe.FindAllStringSubmatch(tableHtml, -1)

		for _, row := range rows {
			rowHtml := row[1]
			thRe := regexp.MustCompile(`<th[^>]*>([\s\S]*?)</th>`)
			tdRe := regexp.MustCompile(`<td[^>]*>([\s\S]*?)</td>`)

			thMatch := thRe.FindStringSubmatch(rowHtml)
			tdMatch := tdRe.FindStringSubmatch(rowHtml)

			if len(thMatch) > 1 && len(tdMatch) > 1 {
				th := strings.TrimSpace(stripTags(thMatch[1]))
				td := strings.TrimSpace(stripTags(tdMatch[1]))

				// ISBN
				if matched, _ := regexp.MatchString(`(?i)isbn`, th); matched {
					isbnRe := regexp.MustCompile(`[\d\-Xx]+`)
					if m := isbnRe.FindStringSubmatch(td); len(m) > 0 {
						detail.ISBN = m[0]
					}
				} else if matched, _ := regexp.MatchString(`出版`, th); matched {
					if matched2, _ := regexp.MatchString(`日期|时间|年`, th); matched2 {
						yearRe := regexp.MustCompile(`(\d{4})`)
						if m := yearRe.FindStringSubmatch(td); len(m) > 1 {
							detail.PublishYear = m[1]
						}
					}
				} else if strings.Contains(th, "索书号") {
					detail.CallNumber = td
				}
			}
		}
	}

	// 默认封面 URL
	if detail.CoverURL == "" {
		detail.CoverURL = fmt.Sprintf("/Cover/Show?instanceId=%s", recordId)
	}

	return detail
}

// ParseHoldings 解析馆藏信息页面
func ParseHoldings(html string, recordId string) []Holding {
	holdings := []Holding{}
	currentLibrary := ""

	// 提取图书馆名称和位置
	h3Re := regexp.MustCompile(`<h3[^>]*>([\s\S]*?)</h3>`)
	h3Matches := h3Re.FindAllStringSubmatchIndex(html, -1)

	type libPosition struct {
		pos     int
		library string
	}
	libPositions := []libPosition{}

	for _, m := range h3Matches {
		text := strings.TrimSpace(stripTags(html[m[2]:m[3]]))
		libMatchRe := regexp.MustCompile(`所属馆[：:]\s*(.+)`)
		if lm := libMatchRe.FindStringSubmatch(text); len(lm) > 1 {
			libPositions = append(libPositions, libPosition{pos: m[0], library: strings.TrimSpace(lm[1])})
		} else if text != "" {
			libPositions = append(libPositions, libPosition{pos: m[0], library: text})
		}
	}

	// 提取表格行
	trRe := regexp.MustCompile(`<tr[^>]*>([\s\S]*?)</tr>`)
	trMatches := trRe.FindAllStringSubmatchIndex(html, -1)

	callnumberRe := regexp.MustCompile(`<span[^>]*class="[^"]*callnumber[^"]*"[^>]*>([\s\S]*?)</span>`)
	barcodeRe := regexp.MustCompile(`<span[^>]*class="[^"]*barcode[^"]*"[^>]*>([\s\S]*?)</span>`)
	availabilityRe := regexp.MustCompile(`<span[^>]*class="[^"]*availability[^"]*"[^>]*>([\s\S]*?)</span>`)

	for _, tr := range trMatches {
		trHtml := html[tr[2]:tr[3]]
		trPos := tr[0]

		// 确定当前所属图书馆
		for _, lp := range libPositions {
			if lp.pos < trPos {
				currentLibrary = lp.library
			}
		}

		callNumber := ""
		barcode := ""
		status := ""

		if m := callnumberRe.FindStringSubmatch(trHtml); len(m) > 1 {
			callNumber = strings.TrimSpace(stripTags(m[1]))
		}
		if m := barcodeRe.FindStringSubmatch(trHtml); len(m) > 1 {
			barcode = strings.TrimSpace(stripTags(m[1]))
		}
		if m := availabilityRe.FindStringSubmatch(trHtml); len(m) > 1 {
			status = strings.TrimSpace(stripTags(m[1]))
		}

		// 如果没有从 class 中提取到状态，尝试从纯文本中提取
		if status == "" {
			statusText := stripTags(trHtml)
			if matched, _ := regexp.MatchString(`已归还|可借|在馆`, statusText); matched {
				sRe := regexp.MustCompile(`(已归还[^\s]*|可借[^\s]*|在馆[^\s]*)`)
				if m := sRe.FindStringSubmatch(statusText); len(m) > 1 {
					status = m[1]
				}
			} else if matched, _ := regexp.MatchString(`借出|外借`, statusText); matched {
				sRe := regexp.MustCompile(`(已借出[^\s]*|外借[^\s]*)`)
				if m := sRe.FindStringSubmatch(statusText); len(m) > 1 {
					status = m[1]
				}
			} else if strings.Contains(statusText, "丢失") {
				status = "馆藏丢失"
			}
		}

		// 翻译英文状态
		status = strings.ReplaceAll(status, "Available", "已归还")
		status = strings.ReplaceAll(status, "Loaned out", "已借出")
		status = strings.ReplaceAll(status, "Lost", "馆藏丢失")

		if callNumber != "" || status != "" {
			holdings = append(holdings, Holding{
				Library:    currentLibrary,
				Location:   barcode,
				CallNumber: callNumber,
				Status:     status,
				RecordID:   recordId,
			})
		}
	}

	return holdings
}
